common/src/main/java/com/google/udmi/util/MessageDowngrader.java

```diff
@@ -103,43 +102,114 @@ private Map<String, Object> downgradeState(JsonNode targetVersion) {
     return JsonUtil.asMap(message);
   }
 
-  private Map<String, Object> downgradeConfig(JsonNode targetVersion) {
+  private Map<String, Object> downgradeConfig(String targetVersion) {
     downgradeConfigRaw(targetVersion);
     return JsonUtil.asMap(message);
   }
 
-  private void downgradeConfigRaw(JsonNode targetVersion) {
-    final String version = convertVersion(targetVersion);
-    String[] components = version.split("-", 2);
-    String[] parts = components[0].split("\\.", 4);
-    major = Integer.parseInt(parts[0]);
-    minor = parts.length >= 2 ? Integer.parseInt(parts[1]) : 0;
-    patch = parts.length >= 3 ? Integer.parseInt(parts[2]) : 0;
+  /**
+   * Returns the current version of the configuration message
+   * @return version
+   */
+  private Integer currentVersion(){
+    return SchemaVersion.fromKey(message.get(VERSION_KEY).asText()).value();
+  }
+
+  private void setCurrentVersion(SchemaVersion version){
+    message.put(VERSION_KEY, version.key());
+  }
+  public boolean wasDowngraded() {
+    return !original.equals(message);
+  }
+
+  private void downgradeConfigRaw(String targetVersionString){
+    final SchemaVersion targetVersionEnum = SchemaVersion.fromKey(targetVersionString);
+    final Integer targetVersion = targetVersionEnum.value();
+
+    // downgrade to 1.4.1
+    if (currentVersion() > VERSION_1_4_1.value() && currentVersion() > targetVersion){
```

Comment:
If we wanted to be really smarmy here we could make a table of sorts that had { versionA, versionB, downgradeFunc, upgradeFun } in it and then all this if statement stuff would be automagic!

Not something I'd suggest for now, but just something to keep in mind for the rainy day in the future when you're bored.